core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -258,15 +258,20 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
     // Make fake resource offers on all executors
     private def makeOffers() {
       // Make sure no executor is killed while some task is launching on it
-      val taskDescs = CoarseGrainedSchedulerBackend.this.synchronized {
-        // Filter out executors under killing
-        val activeExecutors = executorDataMap.filterKeys(executorIsAlive)
-        val workOffers = activeExecutors.map {
-          case (id, executorData) =>
-            new WorkerOffer(id, executorData.executorHost, executorData.freeCores,
-              Some(executorData.executorAddress.hostPort))
-        }.toIndexedSeq
-        scheduler.resourceOffers(workOffers)
+      // SPARK-27112: We need to ensure that there is ordering of lock acquisition
+      // between TaskSchedulerImpl and CoarseGrainedSchedulerBackend objects in order to fix
+      // the deadlock issue exposed in SPARK-27112
+      val taskDescs = scheduler.synchronized {
```

Comment:
I took a quick look at the code that calls this, and I'm wondering if holding the two locks here is really needed.

For context, all this code is inside the RPC endpoint handler. This is a `ThreadSafeRpcEndpoint` so there's only one message being processed at a time, meaning that you won't have multiple threads calling `makeOffers` concurrently.

So it seems to me that it would be possible to:

- with the `CoarseGrainedSchedulerBackend.this` lock held, calculate the works offers.

```
val workOffers = CoarseGrainedSchedulerBackend.this.synchronized {
  ...
}
```

With the scheduler lock held, make the offers:

```
val taskDesc = scheduler.synchronized {
  scheduler.resourceOffers(workOffers)
}
```

And as far as I understand that should work and also be easier to understand, right?

I also noticed that later this code calls `launchTasks`, and that method accesses and modifies data in `executorDataMap` without the `CoarseGrainedSchedulerBackend.this` lock, which is very sketchy.

Reply:
Ok, seems both locks are needed because of SPARK-19757. But the `launchTasks` issue is still there.

Reply:
@vanzin  So in the code, I came across the following comment, wonder if that answers the `launchTasks` issue. I exactly do not understand the intention of the comment though.

```
// Accessing `executorDataMap` in `DriverEndpoint.receive/receiveAndReply` doesn't need any
  // protection. But accessing `executorDataMap` out of `DriverEndpoint.receive/receiveAndReply`
  // must be protected by `CoarseGrainedSchedulerBackend.this`. Besides, `executorDataMap` should
  // only be modified in `DriverEndpoint.receive/receiveAndReply` with protection by
  // `CoarseGrainedSchedulerBackend.this`.
  private val executorDataMap = new HashMap[String, ExecutorData]
```

Reply:
Ok, I think that's fine. I checked and all modifications happen on the endpoint thread, so reading the map from that thread without a lock should be fine. The data being modified (`freeCores`) is also only used in the endpoint thread, so that looks safe too.